copylot/hardware/prime_bsi_camera/camera.py

```diff
@@ -0,0 +1,182 @@
+from pyvcam import pvc, constants
+from pyvcam.camera import Camera
+
+
+class PrimeBSICamera:
+    def __init__(
+        self,
+        camera_name: str = None,
+        # TODO: create properties for these
+        # scan_mode: str = None,
+        # scan_dir: str = None,
+        # binning: tuple = (1, 1),
+        # trigger_mode: str = None,
+        # exposure_out: str = None,
+        enable_metadata: bool = True,
+    ):
+        """
+
+        Parameters
+        ----------
+        camera_name : str
+            Camera name, available through available_cameras().
+            If not provided, the next available camera will be initialized.
+        enable_metadata : bool
+            Toggles camera metadata on and off. By default, it is True.
+        """
+        pvc.init_pvcam()
+
+        try:
+            if camera_name:
+                self.cam = Camera.select_camera(camera_name)
+            else:
+                self.cam = next(Camera.detect_camera())
+        except:
+            pvc.uninit_pvcam()
+            raise
+
+        self.cam.open()
+
+        # Prime BSI Express camera has only one port called 'CMOS' at index 0
+        self.cam.readout_port = 0
+
+        # Hard-coding these values since they won't change
+        # Values can be parsed from self.cam.port_speed_gain_table['CMOS']
+        self._speed_gain_table = {'200 MHz': {'speed_index': 0,
+                                              'Full well': {'gain_index': 1, 'bit_depth': 11},
+                                              'Balanced': {'gain_index': 2, 'bit_depth': 11},
+                                              'Sensitivity': {'gain_index': 3, 'bit_depth': 11}
+                                              },
+                                  '100 MHz': {'speed_index': 1,
+                                              'HDR': {'gain_index': 1, 'bit_depth': 16},
+                                              'CMS': {'gain_index': 2, 'bit_depth': 12}
+                                              }
+                                  }
+
+        self.serial_number = self.cam.serial_no
+        self.sensor_size = self.cam.sensor_size
+        self.chip_name = self.cam.chip_name
+        self.cam.meta_data_enabled = enable_metadata
+
+    def __del__(self):
+        self.cam.finish()
+        self.cam.close()
+        pvc.uninit_pvcam()
+
+    @staticmethod
+    def available_cameras():
+        pvc.init_pvcam()
+        cameras = Camera.get_available_camera_names()
+
+        pvc.uninit_pvcam()
+        return cameras
+
+    @property
+    def speed_gain_table(self):
+        return self._speed_gain_table
```

Comment:
given the `_speed_gain_table` value won't change, I think this can be defined as a class attribute instead of property. What do you think @ieivanov ?

Reply:
Sounds good! By hardcoding this table we're making this class specific to the Prime BSI Express camera, which was our original intention. At a later point we can parse this table from `port_speed_gain_table['CMOS']` to make the class generalize to other PVCAM cameras.

Reply:
absolutely, or we might have separate adapters for the other PVCAM cameras. Either way will be easy to accommodate those.